### PR TITLE
Initial SigNet support

### PR DESCRIPTION
**Pull request by benthecarman**
Closes #1112 

Was able to sync to signet tip from this. To do so I compiled `master` of bitcoin core and synced it on signet, then connected my bitcoin-s instance to it.

```
$ bs-cli getblockcount
4704
```

### CODE REVIEW

**Comment by benthecarman** [comment]
cc @kallewoof would love your review

File: core/src/main/scala/org/bitcoins/core/config/NetworkParameters.scala [hunk 1]
```diff
@@ -203,8 +243,11 @@ object BitcoinNetworks extends StringFactory[BitcoinNetwork] {
...
       TestNet3.p2pkhNetworkByte -> TestNet3,
       TestNet3.p2shNetworkByte -> TestNet3,
-      TestNet3.privateKey -> TestNet3
+      TestNet3.privateKey -> TestNet3,
+      SigNet.p2shNetworkByte -> SigNet,
+      SigNet.p2pkhNetworkByte -> SigNet,
+      SigNet.privateKey -> SigNet
 
-      //ommitting regtest as it has the same network bytes as testnet3
+      //omitting regtest as it has the same network bytes as testnet3
```
Comment (kallewoof) [inline core/src/main/scala/org/bitcoins/core/config/NetworkParameters.scala]: Unless I am mistaken on what this is, signet and testnet share these values as well, just like testnet and regtest.

File: core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala [hunk 2]
```diff
@@ -398,6 +420,80 @@ object RegTestNetChainParams extends BitcoinChainParams {
...
+
+  /** Blocks must satisfy the given script to be considered valid (only for signet networks) */
+  override def signetChallenge: ScriptPubKey = EmptyScriptPubKey
+}
+
+case class SigNetChainParams(
+    signetChallenge: ScriptPubKey = ScriptPubKey.fromAsmHex(
+      "512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be430210359ef5021964fe22d6f8e05b2463c9540ce96883fe3b278760f048f5189f2e6c452ae"))
+    extends BitcoinChainParams {
+  override lazy val networkId = "signet"
```
Comment (kallewoof) [inline core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala:435]: Unusual indentation.

File: core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala [hunk 3]
```diff
@@ -398,6 +420,80 @@ object RegTestNetChainParams extends BitcoinChainParams {
...
+  /**
+    * Minimum amount of chain work on the test network
+    */
+  override lazy val minimumChainWork: BigInteger = {
+    BigInteger.valueOf(0)
+  }
+
+  /**
+    * Regtest allows trivial difficulty blocks
+    * [[https://github.com/bitcoin/bitcoin/blob/006d9ab9e59a00fa02b8d5f32919f9d57567fa65/src/chainparams.cpp#L312 regtest min difficulty]]
```
Comment (kallewoof) [inline core/src/main/scala/org/bitcoins/core/protocol/blockchain/ChainParams.scala]: Nit: I think you mean "Signet does not allow trivial difficulty blocks" and point to the relevant point in bitcoin core here and in below comment on line 485-486.

File: crypto-test/src/test/scala/org/bitcoins/crypto/ECPrivateKeyTest.scala [hunk 4]
```diff
@@ -72,6 +72,8 @@ class ECPrivateKeyTest extends BitcoinSUnitTest {
             assert(ECPrivateKeyUtil.parseNetworkFromWIF(wif).get == network)
           case TestNet3 | RegTest =>
             assert(ECPrivateKeyUtil.parseNetworkFromWIF(wif).get == TestNet3)
+          case SigNet =>
+            assert(ECPrivateKeyUtil.parseNetworkFromWIF(wif).get == SigNet)
```
Comment (kallewoof) [inline crypto-test/src/test/scala/org/bitcoins/crypto/ECPrivateKeyTest.scala]: This can be simplified to `case TestNet3 | Regtest | SigNet =>`, I think?

**Review by kallewoof — COMMENTED** [review]
Looked over code, with some nits that you're welcome to ignore. Don't know scala too well, but looks correct to me.

File: core/src/main/scala/org/bitcoins/core/config/NetworkParameters.scala [hunk 5]
```diff
@@ -243,11 +243,8 @@ object BitcoinNetworks extends StringFactory[BitcoinNetwork] {
...
       TestNet3.p2pkhNetworkByte -> TestNet3,
       TestNet3.p2shNetworkByte -> TestNet3,
-      TestNet3.privateKey -> TestNet3,
-      SigNet.p2shNetworkByte -> SigNet,
-      SigNet.p2pkhNetworkByte -> SigNet,
-      SigNet.privateKey -> SigNet
+      TestNet3.privateKey -> TestNet3
 
-      //omitting regtest as it has the same network bytes as testnet3
+      //omitting regtest and signet as it has the same network bytes as testnet3
```
Comment (kallewoof) [inline core/src/main/scala/org/bitcoins/core/config/NetworkParameters.scala]: Nit: "as they have"

File: core-test/src/test/scala/org/bitcoins/core/protocol/blockchain/ChainParamsTest.scala [hunk 6]
```diff
@@ -101,6 +106,10 @@ class ChainParamsTest extends BitcoinSUnitTest {
         "4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b"))
   }
 
+  it must "generate the correct merkle root for the signet genesis block" in {
+    SigNetChainParams().genesisBlock.blockHeader.merkleRootHashBE.hex must be(
+      "4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b")
```
Comment (nkohen) [inline core-test/src/test/scala/org/bitcoins/core/protocol/blockchain/ChainParamsTest.scala:107]: should probably assert this for a couple different challenge spks

**Review by nkohen — APPROVED** [review]
Looks good

**Review by rorp — APPROVED** [review]

File: core/src/main/scala/org/bitcoins/core/protocol/BtcHumanReadablePart.scala [hunk 7]
```diff
@@ -53,6 +53,7 @@ object BtcHumanReadablePart extends StringFactory[BtcHumanReadablePart] {
       case _: MainNet  => bc
       case _: TestNet3 => tb
       case _: RegTest  => bcrt
+      case _: SigNet   => tb
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/protocol/BtcHumanReadablePart.scala:56]: :-(, they didn't give signet a unique prefix ?

**Review by Christewart — APPROVED** [review]

File: core/src/main/scala/org/bitcoins/core/protocol/BtcHumanReadablePart.scala (hunk 7 above)
Comment (kallewoof) [reply core/src/main/scala/org/bitcoins/core/protocol/BtcHumanReadablePart.scala:56]: Initially we did, but apparently some people want all test nets to use the same prefix. Even regtest and testnet.

**Pull request merged**